XiEditor/Trace.swift

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A simple mechanism for logging trace events and outputting a file in
+// Chrome tracing format
+
+import Foundation
+
+var globalTrace = Trace()
+
+class Trace {
+    let mutex = UnfairLock()
+    let BUF_SIZE = 100_000
+    var buf: [TraceEntry]
```

Comment:
This might make more sense as a `TraceEntry?`, which would mean we wouldn't need to have the empty init method on `TraceEntry` that just creates a dummy entry; and it would also clean up the logic in `write` where you're doing bounds checking against the number of entries.

Reply:
It doesn't clean up `write`, as that's doing circular buffer logic and not simply bounds-checking.